src/java.base/share/classes/java/security/Provider.java

```diff
@@ -1149,7 +1149,7 @@ private void parseLegacy(String name, String value, OPType opType) {
             switch (opType) {
                 case ADD:
                     // clean up old alias if present
-                    Service prevAliasService = legacyMap.get(aliasAlg);
+                    Service prevAliasService = legacyMap.get(aliasKey);
```

Comment:
The change looks okay but I'm surprised this wasn't caught by tests. 
@valeriepeng - would it be feasible to have a test that removes an old alias so that this code is exercised by tests?

Reply:
Hmm, this "ADD" case should be covered by existing regression tests. I will take a look. Thanks.

Reply:
Hmm, existing test does add an alias, the legacyMap.get() would returns null upon 'aliasAlg' (type: String). To test this particular scenario, I enhanced existing regression to re-assign an alias and then check if the returned impl is as expected. Here is the proposed test patch:
`
diff --git a/test/jdk/java/security/Provider/CaseSensitiveServices.java b/test/jdk/java/security/Provider/CaseSensitiveServices.java
index 2a6676b4f58..5b0e0f1fb0c 100644
--- a/test/jdk/java/security/Provider/CaseSensitiveServices.java
+++ b/test/jdk/java/security/Provider/CaseSensitiveServices.java
@@ -36,9 +36,12 @@ public class CaseSensitiveServices extends Provider {
         super("Foo", "1.0", null);
         put("MessageDigest.Foo", "com.Foo");
         put("mESSAGEdIGEST.fOO xYz", "aBc");
-        put("ALg.aliaS.MESSAGEdigest.Fu", "FoO");
+        // first assign the DEF alias to algorithm Foo
+        put("ALg.aliaS.MESSAGEdigest.DEF", "FoO");
         put("messageDigest.Bar", "com.Bar");
         put("MESSAGEDIGEST.BAZ", "com.Baz");
+        // reassign the DEF alias to algorithm Bar
+        put("ALg.aliaS.MESSAGEdigest.DEF", "Bar");
     }

     public static void main(String[] args) throws Exception {
@@ -47,12 +50,18 @@ public class CaseSensitiveServices extends Provider {
         if (p.getServices().size() != 3) {
             throw new Exception("services.size() should be 3");
         }
+
         Service s = testService(p, "MessageDigest", "fOO");
         String val = s.getAttribute("Xyz");
         if ("aBc".equals(val) == false) {
             throw new Exception("Wrong value: " + val);
         }
-        testService(p, "MessageDigest", "fU");
+
+        // test Service alias DEF and its associated impl is Bar
+        s = testService(p, "MessageDigest", "DeF");
+        if (s.getAttribute("Xyz") != null) {
+            throw new Exception("DEF mapped to the wrong impl");
+        }
         testService(p, "MessageDigest", "BAR");
         testService(p, "MessageDigest", "baz");
         System.out.println("OK");
@@ -67,5 +76,4 @@ public class CaseSensitiveServices extends Provider {
         }
         return s;
     }
-
 }
`

Reply:
Thanks, @valeriepeng! Do you see the current JDK 18 failing with that test addition? Because this still passes for me without the fix from this PR: `CONF=linux-x86_64-server-fastdebug make run-test TEST=java/security/Provider/CaseSensitiveServices.java`

Reply:
Ping @valeriepeng ^^^

Reply:
The test update checks that the alias maps to the right algorithm. However, the particular code change covered by this fix is for removing the alias from the previous algorithm which is not triggered by the test change. Alias is an internal property of the Service object. Let me explore more and see if it can be triggered.

Reply:
There is no public API for retrieving the aliases from a given Service object. On it's reported as part of toString(). So, we can also check the toString() output to see if the alias is removed from the previous service. Updated test patch attached
[patch.txt](https://github.com/openjdk/jdk18/files/7804357/patch.txt)
.

Reply:
All right, thank you, @valeriepeng! The new test fails without the patch, and passes with it. I added it to PR.